athena-elasticsearch/src/main/java/com/amazonaws/connectors/athena/elasticsearch/AwsRestHighLevelClientFactory.java

```diff
@@ -33,38 +35,71 @@
 {
     private static final Logger logger = LoggerFactory.getLogger(AwsRestHighLevelClientFactory.class);
 
-    private final boolean useAwsCredentials;
-
     /**
      * credentialsPattern is used for parsing out the username/password credentials out of the endpoint with the
      * format: "username@password:" (e.g. if the endpoint is http://myusername@mypassword:www.google.com, then this
      * pattern will be used to extract the following credentials string: "myusername@mypassword:").
      */
     private static final Pattern credentialsPattern = Pattern.compile("[^/@]+@[^:]+:");
 
+    /**
+     * A flag indicating how the created clients will authenticate to an Elasticsearch instance.
+     * If useAwsCredentials = true, the client will be injected with AWS credentials.
+     * If useAwsCredentials = false and username/password are extracted using the credentialsPattern, the client
+     * will be injected with username/password credentials.
+     * If useAwsCredentials = false and username/password are not present, the client will be created with no
+     * credentials.
+     */
+    private final boolean useAwsCredentials;
+
+    /**
+     * Stored cache of clients accessible via the endpoint.
+     */
+    private final Map<String, AwsRestHighLevelClient> clientCache;
+
     /**
      * Constructs a new client factory (using a builder) that will create clients injected with credentials.
      * @param useAwsCredentials if true the factory create clients injected with AWC credentials.
-     *                              If false, it will create clients injected with username/password credentials.
+     *                          If false, it will create clients injected with username/password credentials.
      */
-    protected AwsRestHighLevelClientFactory(boolean useAwsCredentials)
+    public AwsRestHighLevelClientFactory(boolean useAwsCredentials)
     {
         this.useAwsCredentials = useAwsCredentials;
+        this.clientCache = new HashMap<>();
+    }
+
+    /**
+     * Attempts to get a REST client from the client cache synchronously using the endpoint as a key. If the client
+     * does not exist in the cache, createClient() is called to create a new client. The newly created client is
+     * then inserted into the cache and stored for later use.
+     * @param endpoint is the Elasticsearch instance endpoint.
+     * @return an Elasticsearch REST client.
+     */
+    public synchronized AwsRestHighLevelClient getClient(String endpoint)
```

Comment:
rename to getOrCreate(...) which is a better description of what it does.

Reply:
changed to getOrCreate(...)